### PR TITLE
drivers: bbled: pwm: mchp: Enable BBLED low power feature

### PR DESCRIPTION
**Pull request by Manimaran-A**
Updated the driver to support low power mode.
Introduced "enable-low-power" flag in device tree to control(on/off) low power mode.

If flag added in DTS, during sleep BBLED will switch off the LEDs. Otherwise BBLED will continue the configured blinking pattern on LEDs.

### CODE REVIEW

**Review by jvasanth1 — CHANGES_REQUESTED** [review]

File: drivers/pwm/pwm_mchp_xec_bbled.c
Comment (Manimaran-A) [reply drivers/pwm/pwm_mchp_xec_bbled.c]: For 48MHz clock, logic updated to execute PM_SUSPEND and PM_RESUME blocks.

**Review by jvasanth1 — DISMISSED** [review]

File: dts/bindings/pwm/microchip,xec-pwmbbled.yaml
Comment (henrikbrixandersen) [inline dts/bindings/pwm/microchip,xec-pwmbbled.yaml:54]: Can we improve the description a bit? From a user point of view, I find it very difficult to parse the relationship between this property, the power management subsystem, and the input clock.

File: drivers/pwm/pwm_mchp_xec_bbled.c
Comment (henrikbrixandersen) [inline drivers/pwm/pwm_mchp_xec_bbled.c:429]: Boolean DTS properties are always defined, so using `_OR()` doesn't make sense here. Please just store the boolean in a `bool` variable without using `_OR()`.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: dts/bindings/pwm/microchip,xec-pwmbbled.yaml
Comment (Manimaran-A) [reply dts/bindings/pwm/microchip,xec-pwmbbled.yaml:54]: @henrikbrixandersen Is below ok?

BBLED has two clock inputs.
 1. Main system clock (48MHz)
 2. 32KHz Core clock  (32.768KHz)
When BBLED enter into Suspend state, 48MHz clock
will be switched off by PCR(Power, Clock and Reset) block.
But 32KHz Core clock will be available to BBLED.

There may be a product requirement, either to blink (or) not blink LED
in Suspend state.

"enable-low-power-32k" flag shall be used along with 32KHz clock to blink (or) not blink the LED in Suspend state.

Comment (henrikbrixandersen) [reply dts/bindings/pwm/microchip,xec-pwmbbled.yaml:54]: Yes, if you could update the description to something along those lines, I think will be a huge improvement.

**Comment by henrikbrixandersen** [comment]
Related: #58466

**Comment by simonguinot** [comment]
> Related: #58466

Thanks @henrikbrixandersen.

Hi @jvasanth1 and @Manimaran-A. I am quite confused here. Please can you explain the relationship between the `led_mchp_xec` and `pwm_mchp_xec_bbled` drivers ? It looks like they're running the same device but I can't find any relationship or dependencies between them or their DT bindings. In addition I can't find any users for the `microchip,xec-pwmbbled` compatible in the Zephyr tree.

**Comment by Manimaran-A** [comment]
@simonguinot 
Both drivers are using the BBLED block. They don't have dependency between them.
"led_mchp_xec" driver implemented to support zephyr LED API (on, off and blink).
"pwm_mchp_xec_bbled" driver implemented to support zephyr PWM API (set_cycles and get_cycles_per_sec).

**Comment by Manimaran-A** [comment]
Yes, led_mchp_xec driver is simple and configured to use Zephyr LED API with 32KHz.

And some of our customer wants to access the BBLED block using Zephyr LED API, and some others want to have through Zephyr PWM API. So created two different driver and one of them can be configured based on the use case.

**Comment by Manimaran-A** [comment]
we are not using led-pwm driver on top of bbled pwm driver; can you clarify that, either one of them can be configured based on zephyr API use case

**Comment by simonguinot** [comment]
> we are not using led-pwm driver on top of bbled pwm driver; can you clarify that, either one of them can be configured based on zephyr API use case

The `led-pwm` driver don't configure the PWM directly but relies on the PWM API (`set_cycles` method). Please have a look at it.

Here is a simplified example of DT:
```
pwmleds {
        compatible = "pwm-leds";
        bbled0 {
                pwms = <&bbled_pwm0 ...>;
        };      
        bbled1 {
                pwms = <&bbled_pwm1 ...>;
        };
       ...
};      

bbled_pwm0 {
        compatible = "microchip,xec-pwmbbled";
        ...
};
bbled_pwm1 {
        compatible = "microchip,xec-pwmbbled";
        ...
};
...
```
The `led-pwm` driver is the standard choice when a LED is connected to a PWM. I am probably missing something obvious here but I don't see why we need a specific BBLED LED driver. Please fix me :)

**Comment by scottwcpg** [comment]
@simonguinot MCHP has one customer demanding PWM API for the BBLED HW and another customer demanding the LED API for BBLED HW.  There will be both PWM and LED drivers. Is it worth re-working both drivers so the LED driver makes calls to the PWM driver?

**Comment by simonguinot** [comment]
> 

Hi @scottwcpg,

I think that the `led_mchp_xec` driver could just be removed. The customer requiring the PWM API would have it with the `pwm_mchp_xec_bbled` driver. And the customer requiring the LED API would have it with the `led_pwm` driver. That's it.

A specific BBLED LED driver only makes sense if the `led-pwm` driver can't be used (not compatible, missing features, etc).

Please can you check how the `led-pwm` driver works for you ?

I apologize for not raising this point when reviewing the initial PR of the BBLED LED driver. For some reason I missed the PR of the BBLED PWM driver...

**Comment by albertofloyd** [comment]
> `led_pwm`

led_pwm generic  driver has no 

> > 
> 
> Hi @scottwcpg,
> 
> I think that the `led_mchp_xec` driver could just be removed. The customer requiring the PWM API would have it with the `pwm_mchp_xec_bbled` driver. And the customer requiring the LED API would have it with the `led_pwm` driver. That's it.
> 
> A specific BBLED LED driver only makes sense if the `led-pwm` driver can't be used (not compatible, missing features, etc).
> 
> Please can you check how the `led-pwm` driver works for you ?
> 
> I apologize for not raising this point when reviewing the initial PR of the BBLED LED driver. For some reason I missed the PR of the BBLED PWM driver...

@simonguinot  Agree to generic API point. However, from what I see the generic led-pwm driver assumes the underlying HW block is PWM for device power management purposes. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led/led_pwm.c#L105

Which for MCHP once PR https://github.com/zephyrproject-rtos/zephyr/pull/60051 is merged, it only changes the pin.

However, the LED BBHW while it can be used as PWM, it has an additional power feature option (ability to run from other than main clock) that cannot be taken advantage with said driver

**Comment by simonguinot** [comment]
> @simonguinot Agree to generic API point. However, from what I see the generic led-pwm driver assumes the underlying HW block is PWM for device power management purposes. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led/led_pwm.c#L105
> 
> Which for MCHP once PR #60051 is merged, it only changes the pin.
> 
> However, the LED BBHW while it can be used as PWM, it has an additional power feature option (ability to run from other than main clock) that cannot be taken advantage with said driver

Hi @albertofloyd,

Yes the `led-pwm` driver relies on PWM for power management. It calls `pm_device_action_run()` for the underlying PWM device. Here the PWM driver will be `pwm_mchp_xec_bbled`. So at then end it is the `xec_bbled_pm_action` function which will do the work once #60051 will be merged.

Honestly the code proposed for `pwm_xec_pm_action` and `xec_bbled_pm_action` looks very similar:

- https://github.com/zephyrproject-rtos/zephyr/pull/60051/commits/9149319b37d3524e3b3588ea7fd16c1a85ce1547
- https://github.com/zephyrproject-rtos/zephyr/pull/58466/commits/6ecb57d86423cd0c0384d42ed1473e4fe3403d69

So PM support should be just fine. In addition the BBLED PWM driver should take in charge the clock configuration as well if needed.

Am I missing something ?

Let's wait for an answer from Microchip. But the more I am looking at it the less I understand why we introduced a specific LED driver in a first place.

**Comment by scottwcpg** [comment]
@simonguinot @albertofloyd I will do an experiment with pwm-led to find out if it works.  DT needs to pass the flag to select 32KHz as the BBLED-PWM driver clock when using pwm-led.

File: dts/bindings/pwm/microchip,xec-pwmbbled.yaml
Comment (Manimaran-A) [reply dts/bindings/pwm/microchip,xec-pwmbbled.yaml]: Updated the description

File: drivers/pwm/pwm_mchp_xec_bbled.c
Comment (Manimaran-A) [reply drivers/pwm/pwm_mchp_xec_bbled.c]: Updated with DT_INST_PROP macro

**Comment by scottwcpg** [comment]
@simonguinot @albertofloyd Did the experiment using LED-PWM driver with BBLED HW.  Two issues found:
1. The LED-PWM sample fails when using the set brightness API.  The root cause is MCHP PWM-BBLED driver reports cycles per second value incorrectly. The BBLED block uses either 32KHz or 48MHz input but has a built-in divider of 256 so the actual maximum frequencies it can do are 32K/256 and 48M/256. Changed driver to report actual cycles per sec and encountered second issue.
2. Bug in led_pwm.c exposed by setting the pwm_dt_spec period parameter to 1 second. The DT period parameter is in nanoseconds and is a uint32_t.  A value of 1 second does fit in 32-bits.  led_pwm_set_brightness() calculates the pulse cycles by dt_led->period * value / 100 where value = 100 for full ON.  This calculation overflows 32-bits resulting in truncation of the value passed to the MCHP PWM-BBLED driver.  You don't get full ON instead you get a small duty cycle.  I interpret the period parameter of pwm_dt_spec as the period you want the LED to blink that is why I set it 1 second.  Observed some other users of pwm_dt_spec using large values.

Do you agree this is a bug in led_pwm.c ?

**Comment by simonguinot** [comment]
> @simonguinot @albertofloyd Did the experiment using LED-PWM driver with BBLED HW. Two issues found:

If we put those bugs aside, are you happy with the `led_pwm` driver ? Is it a valid replacement for the `led_mchp_xec` LED driver ? I think you gain as a bonus the `set_brightness` API :)

> 
>     1. The LED-PWM sample fails when using the set brightness API.  The root cause is MCHP PWM-BBLED driver reports cycles per second value incorrectly. The BBLED block uses either 32KHz or 48MHz input but has a built-in divider of 256 so the actual maximum frequencies it can do are 32K/256 and 48M/256. Changed driver to report actual cycles per sec and encountered second issue.

OK.

> 
>     2. Bug in led_pwm.c exposed by setting the pwm_dt_spec period parameter to 1 second. The DT period parameter is in nanoseconds and is a uint32_t.  A value of 1 second does fit in 32-bits.  led_pwm_set_brightness() calculates the pulse cycles by dt_led->period * value / 100 where value = 100 for full ON.  This calculation overflows 32-bits resulting in truncation of the value passed to the MCHP PWM-BBLED driver.  You don't get full ON instead you get a small duty cycle.  I interpret the period parameter of pwm_dt_spec as the period you want the LED to blink that is why I set it 1 second.  Observed some other users of pwm_dt_spec using large values.
> 
> 
> Do you agree this is a bug in led_pwm.c ?

I looked at the code and I agree that there is indeed an overflow bug in the `led-pwm` driver if the PWM period is too large. 
But a 1 second PWM period doesn't make much sense. It is way too big ! A period greater than 50/100 ms won't be able to fool human eye for full brightness. The eye will perceive the blinking.

AFAIK this period is only used for the `set_brightness` function. The `blink` function calls `pwm_set_dt` which basically uses a custom period computed and passed by the `led-pwm` driver. The period in `pwm_dt_spec` is ignored.

So to have a full brightness granularity you want at least a 255 ns period. And as I said above, it shouldn't be more than 50-100ms. Any values between should work. IMHO a 10 or 20 ms default period would be a good default for a PWM LED in DT .

**Comment by scottwcpg** [comment]
@simonguinot @albertofloyd I agree the PWM-LED can be used and we can eliminate the MCHP BBLED driver.  I looked through the Zephyr pwd_dt_spec online documentation and pwm.h.  The documentation and comments only state "PWM signal period in nanoseconds".  What should this DT value represent. The maximum PWM frequency of the specific PWM controller?

**Comment by simonguinot** [comment]
> @simonguinot @albertofloyd I agree the PWM-LED can be used and we can eliminate the MCHP BBLED driver. I looked through the Zephyr pwd_dt_spec online documentation and pwm.h. The documentation and comments only state "PWM signal period in nanoseconds". What should this DT value represent. The maximum PWM frequency of the specific PWM controller?

For the `led-pwm` driver it is the default PWM period. And it is used by the `set_brightness` API. See my previous comment. 10 or 20 milliseconds is good choice. I agree it should be documented / explained somewhere. A period of 1 second makes no sense here.

File: drivers/pwm/pwm_mchp_xec_bbled.c
Comment (henrikbrixandersen) [inline drivers/pwm/pwm_mchp_xec_bbled.c:415]: `PM_DEVICE_DT_INST_DEFINE()` does not need to be wrapped in `#fdef CONFIG_PM_DEVICE`.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/pwm/pwm_mchp_xec_bbled.c
Comment (Manimaran-A) [reply drivers/pwm/pwm_mchp_xec_bbled.c]: Ok, updated.

**Review by albertofloyd — APPROVED** [review]

**Comment by Manimaran-A** [comment]
@henrikbrixandersen could you please help to review and approve?

File: dts/bindings/pwm/microchip,xec-pwmbbled.yaml
Comment (henrikbrixandersen) [inline dts/bindings/pwm/microchip,xec-pwmbbled.yaml:58]: It's a property, not a flag.

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**
